text/0008-kadalu-rebalance.adoc

```diff
@@ -0,0 +1,132 @@
+	--xlator-option "replicate.entry-self-heal=off"                          \
+	--xlator-option "*distribute.readdir-optimize=on"                        \
+	--xlator-option "*distribute.rebalance-cmd=<RebalanceCmd.FixLayoutStart> \
+	--xlator-option "*distribute.node-uuid=<node-id>"                        \
+        --xlator-option "*distribute.commit-hash=<unique-id>"                    \
+        -p <pid-file>                                                            \
+        --socket-file <socket-file>                                              \
+        -l <logfile>
+----
+
+== Fix layout Rebalance (Operator based)
+
+Operator parses the CRD modification and understands that Storage Pool expansion is triggered. After starting all the new Server pods. Start a Job that triggers the Fix layout rebalance. kubectl-kadalu can get the status of rebalance by checking the status of the Job.
+
+Refer the above example command to be run as job.
+
+== Files rebalance (Moana based)
+
+This is the manual process can be triggered by admins based on the needs (When no space available in the existing Storage units, or when the Cluster usage is minimal).
+
+"Volume expand" command also prints the note with the steps to run the Rebalance command. Rebalance helps to distribute the existing files uniformly among the Storage units.
+
+----
+kadalu volume rebalance-start <POOL>/<VOLUME>
+----
+
+Stop and Start command will start the rebalance process from the beginning, it doesn't remember the previous progress.
+
+----
+kadalu volume rebalance-stop <POOL>/<VOLUME>
+----
+
+Above command calls the respective API that internally starts a service in respective Storage unit nodes. **Note**: These services will halt once the Rebalance process completes its job. Do not start the service in every nodes of the Volume, start only in the node of the first Storage unit from each distribute groups.
+
+Service in each node will mount the Volume locally and runs the below command (More info related to this sub-command is covered in the next section)
+
+----
+kadalu rebalance-process <storage-unit-path> <mount-path>
+----
+
+
+== Files Rebalance (Operator based)
+
+Introduce a new CRD to start the rebalance process.
+
+[source,yaml]
+----
+# File: rebalance-storage-pool-1.yaml
+---
+apiVersion: kadalu-operator.storage/v1alpha1
+kind: KadaluStorageRebalance
+metadata:
+  # This will be used as Job name
+  name: rebal-sp1-dec-2022
+spec:
+  pool: storage-pool1
+----
+
+Operator gets the Storage pool info from configmap and then identifies the first Storage unit(Brick) from each distribute groups. Based on the node affinity of each Storage unit, Operator starts the Rebalance job (Refer next section).
+
+Delete this CRD to stop the Rebalance job.
+
+`kubectl-kadalu` will collect the status from each Jobs and shows the aggregated status.
+
+----
+kubectl kadalu rebalance-status <Pool-name>
+----
+
+== Tool to handle the rebalance
```

Comment:
Does this also contain the migrate data option to remove linkfiles created for renamed files in cached locations or this is part of glusterfs rebalance command itself?

Reply:
Migrate data is handled by GlusterFS itself via the exposed virtual xattrs. Rebalance tool only manages the filesystem crawl and calls Rebalance xattrs.